CADability/ModelView.cs

```diff
@@ -758,69 +759,74 @@ bool ICommandHandler.OnUpdateCommand(string MenuId, CommandState CommandState)
         void IView.OnPaint(PaintEventArgs e)
         {
             if (projectedModel == null || Projection == null) return; // should never haappen
-            IPaintTo3D paintTo3D = canvas.PaintTo3D;
-            if (paintTo3D != null)
-            {
-                paintTo3D.MakeCurrent();
-                paintTo3D.Clear(BackgroundColor);
-                PaintClearEvent?.Invoke(e.ClipRectangle, this, paintTo3D);
-                paintTo3D.UseZBuffer(true);
-                BoundingCube bc = Model.Extent;
-                bc.MinMax(Model.MinExtend);
-                // sicherstellen, dass die komplette Rasterebene auch mit angezeigt wird
-                BoundingRect ext = BoundingRect.EmptyBoundingRect;
-                ext.MinMax(Projection.DrawingPlane.Project(new GeoPoint(bc.Xmin, bc.Ymin, bc.Zmin)));
-                ext.MinMax(Projection.DrawingPlane.Project(new GeoPoint(bc.Xmin, bc.Ymin, bc.Zmax)));
-                ext.MinMax(Projection.DrawingPlane.Project(new GeoPoint(bc.Xmin, bc.Ymax, bc.Zmin)));
-                ext.MinMax(Projection.DrawingPlane.Project(new GeoPoint(bc.Xmin, bc.Ymax, bc.Zmax)));
-                ext.MinMax(Projection.DrawingPlane.Project(new GeoPoint(bc.Xmax, bc.Ymin, bc.Zmin)));
-                ext.MinMax(Projection.DrawingPlane.Project(new GeoPoint(bc.Xmax, bc.Ymin, bc.Zmax)));
-                ext.MinMax(Projection.DrawingPlane.Project(new GeoPoint(bc.Xmax, bc.Ymax, bc.Zmin)));
-                ext.MinMax(Projection.DrawingPlane.Project(new GeoPoint(bc.Xmax, bc.Ymax, bc.Zmax)));
-                ext.Inflate(ext.Width, ext.Height);
-                bc.MinMax(Projection.DrawingPlane.ToGlobal(new GeoPoint2D(ext.Left, ext.Bottom)));
-                bc.MinMax(Projection.DrawingPlane.ToGlobal(new GeoPoint2D(ext.Left, ext.Top)));
-                bc.MinMax(Projection.DrawingPlane.ToGlobal(new GeoPoint2D(ext.Right, ext.Bottom)));
-                bc.MinMax(Projection.DrawingPlane.ToGlobal(new GeoPoint2D(ext.Right, ext.Top)));
-
-                if (!additionalExtent.IsEmpty) bc.MinMax(additionalExtent);
-
-                paintTo3D.SetProjection(Projection, bc);
-
-                double precisionFactor = Settings.GlobalSettings.GetDoubleValue("HighestDisplayPrecision", 1e5);
-                // HighestDisplayPrecision ist nicht interaktiv einstellbar, nur mit Programmcode
-                // Wenn man weiter hineinzoomt, wird keine weitere Displayliste mehr berechnet
-                // Sollte man dokumentieren
-                if (displayPrecision > 0) paintTo3D.Precision = Math.Min(displayPrecision, Math.Max(1.0 / Projection.WorldToDeviceFactor, Model.Extent.MaxSide / precisionFactor));
-                else paintTo3D.Precision = Math.Max(1.0 / Projection.WorldToDeviceFactor, Model.Extent.MaxSide / precisionFactor);
-
-                try
+            if (!painting)
```

Comment:
Personally I try to avoid mixing program logic and reentrancy checking. That keeps the code cleaner and makes review easier because if you don't have to wrap your existing code in a large if block, you don't get so many changes in a PR.

(Just my thoughts, I'm not a maintainer)

```
private bool painting = false; //True during OnPaint. To avoid reentrancy.
void IView.OnPaint(PaintEventArgs e)
{
    if(painting)
    {
        return;
    }
    try
    {
        painting = true;
        PerformPaint(e);
    }
    finally
    {
        painting = false;
    }
}

private void PerformPaint(PaintEventArgs e)
{
    // initial logic goes here
}
```

Reply:
Good Idea to avoid logic and reentrancy checking.
I personally don't like return not at the end of the methods, because at the end they will be used to many and the code will be not easy to understand.
Regarding try/catch, I think that should be checked where the problems can occur.

Reply:
> Regarding try/catch, I think that should be checked where the problems can occur

I agree on this one. I only used a finally block which will execute even if the code in the try block raises an exception. Otherwise the variable `painting` would be false forever after an exception.